File: Server~/src/tools/gameobject/DeleteGameObjectTool.ts
import { z } from 'zod';
import { BaseTool } from '../base/BaseTool.js';
import { Tool, Tags } from '../base/ToolDecorators.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

/**
 * Input schema for delete_gameobject tool
 * Based on Unity API: Undo.DestroyObjectImmediate
 * https://docs.unity3d.com/ScriptReference/Undo.DestroyObjectImmediate.html
 */
const DeleteGameObjectToolArgsSchema = z.object({
  instanceIds: z.array(z.number().int()).optional().describe('Array of GameObject instance IDs to delete'),
  objectPath: z.string().optional().describe('Hierarchy path to the GameObject to delete. Example: "Parent/Child"'),
  objectName: z.string().optional().describe('Name of the GameObject(s) to delete. Deletes all matching objects.')
});

/**
 * DeleteGameObject Tool
 * Deletes GameObjects from the scene with Undo support
 */
@Tool({
  name: 'delete_gameobject',
  description: 'Delete GameObjects from the scene. Supports Undo/Redo.',
  category: 'gameobject',
  version: '1.0.0'
})
@Tags(['unity', 'gameobject', 'delete', 'destroy'])
export class DeleteGameObjectTool extends BaseTool {
  get name() {
    return 'delete_gameobject';
  }

  get description() {
    return 'Delete GameObjects from the scene. Supports Undo/Redo. Use Edit > Undo to restore deleted objects.';
  }

  get inputSchema() {
    return DeleteGameObjectToolArgsSchema;
  }

  get category() {
    return 'gameobject';
  }

  protected formatSuccessResponse(result: any): CallToolResult {
    return {
      content: [{
        type: 'text',
        text: `✅ ${result.message || 'GameObjects deleted successfully'}`
      }]
    };
  }
}

File: Editor/Tools/GameObject/DeleteGameObjectTool.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;
using McpUnity.Utils;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool to delete GameObjects from the scene
    /// Based on Unity API: Undo.DestroyObjectImmediate
    /// https://docs.unity3d.com/ScriptReference/Undo.DestroyObjectImmediate.html
    /// </summary>
    public class DeleteGameObjectTool : McpToolBase
    {
        public DeleteGameObjectTool()
        {
            Name = "delete_gameobject";
            Description = "Delete GameObjects from the scene. Supports Undo/Redo.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                JArray instanceIdsArray = parameters["instanceIds"] as JArray;
                string objectPath = parameters["objectPath"]?.ToObject<string>();
                string objectName = parameters["objectName"]?.ToObject<string>();
                bool includeChildren = parameters["includeChildren"]?.ToObject<bool>() ?? true;

                List<GameObject> objectsToDelete = new List<GameObject>();

                // Method 1: By instance IDs
                if (instanceIdsArray != null && instanceIdsArray.Count > 0)
                {
                    foreach (var id in instanceIdsArray)
                    {
                        GameObject obj = EditorUtility.InstanceIDToObject(id.ToObject<int>()) as GameObject;
                        if (obj != null)
                        {
                            objectsToDelete.Add(obj);
                        }
                    }
                }

                // Method 2: By object path
                if (!string.IsNullOrEmpty(objectPath))
                {
                    GameObject obj = GameObject.Find(objectPath);
                    if (obj != null && !objectsToDelete.Contains(obj))
                    {
                        objectsToDelete.Add(obj);
                    }
                }

                // Method 3: By object name (finds all with that name)
                if (!string.IsNullOrEmpty(objectName))
                {
                    GameObject[] allObjects = UnityEngine.Object.FindObjectsByType<GameObject>(FindObjectsSortMode.None);
                    foreach (var obj in allObjects)
                    {
                        if (obj.name == objectName && !objectsToDelete.Contains(obj))
                        {
                            objectsToDelete.Add(obj);
                        }
                    }
                }

                if (objectsToDelete.Count == 0)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "No GameObjects found to delete. Provide 'instanceIds', 'objectPath', or 'objectName'.",
                        "validation_error");
                }

                JArray deletedArray = new JArray();
                int count = 0;

                foreach (GameObject obj in objectsToDelete)
                {
                    string deletedName = obj.name;
                    int deletedId = obj.GetInstanceID();
                    int childCount = obj.transform.childCount;

                    // Use Undo.DestroyObjectImmediate for undo support
                    Undo.DestroyObjectImmediate(obj);

                    deletedArray.Add(new JObject
                    {
                        ["name"] = deletedName,
                        ["instanceId"] = deletedId,
                        ["childCount"] = includeChildren ? childCount : 0
                    });
                    count++;
                }

                return new JObject
                {
                    ["success"] = true,
                    ["message"] = $"Deleted {count} GameObject(s). Use Edit > Undo to restore.",
                    ["count"] = count,
                    ["deletedObjects"] = deletedArray
                };
            }
            catch (Exception ex)
            {
                McpLogger.LogError($"DeleteGameObjectTool error: {ex.Message}");
                return McpUnitySocketHandler.CreateErrorResponse($"Failed: {ex.Message}", "execution_error");
            }
        }
    }
}
